### PR TITLE
Add disjoint RefInput PredicateFailure

### PR DESCRIPTION
**Pull request by TimSheard**
Address issue #4007 
Added some unit tests in Test.Cardano.Ledger.Example.BabbageFeatures.hs
Generalized the BabbageFeatures to work on Babbage and Conway Eras.

# Checklist

- [ ] Commit sequence broadly makes sense and commits have useful messages
- [ ] New tests are added if needed and existing tests are updated
- [ ] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [ ] The version bounds in `.cabal` files for all affected packages are updated. **If you change the bounds in a cabal file, that package itself must have a version increase.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [ ] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages. **New section is never added with the code changes.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/intersectmbo/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [x] Self-reviewed the diff

### CODE REVIEW

File: eras/babbage/impl/src/Cardano/Ledger/Babbage/Rules/Utxo.hs [hunk 1]
```diff
@@ -320,6 +342,13 @@ utxoTransition = do
   {-   txb := txbody tx   -}
   let txBody = body tx
       allInputs = txBody ^. allInputsTxBodyF
+      refInputs :: Set (TxIn (EraCrypto era))
+      refInputs = txBody ^. referenceInputsTxBodyL
+      inputs :: Set (TxIn (EraCrypto era))
+      inputs = txBody ^. inputsTxBodyL
+
+  {- inputs ∩ refInputs = ∅ -}
+  runTest $ disjointRefInputs @era pp inputs refInputs
```
Comment (lehins) [inline eras/babbage/impl/src/Cardano/Ledger/Babbage/Rules/Utxo.hs]: PParams is injective, so there is no need to pass `era` type parameter
```suggestion
  runTest $ disjointRefInputs pp inputs refInputs
```

File: eras/babbage/impl/src/Cardano/Ledger/Babbage/Rules/Utxo.hs [hunk 2]
```diff
@@ -194,6 +202,20 @@ feesOK pp tx (UTxO utxo) =
...
+  forall era.
+  EraPParams era =>
+  PParams era ->
+  Set (TxIn (EraCrypto era)) ->
+  Set (TxIn (EraCrypto era)) ->
+  Test (BabbageUtxoPredFailure era)
+disjointRefInputs pp inputs refInputs =
+  when
+    (pvMajor (pp ^. ppProtocolVersionL) > eraProtVerHigh @(BabbageEra (EraCrypto era)))
+    (failureIf (null common) (BabbageNonDisjointRefInputs common))
```
Comment (lehins) [inline eras/babbage/impl/src/Cardano/Ledger/Babbage/Rules/Utxo.hs]: Let's switch `BabbageNonDisjointRefInputs` to `NonEmpty` from a Set. 
```suggestion
    (failureOnNonEmpty common BabbageNonDisjointRefInputs)
```

File: eras/babbage/impl/src/Cardano/Ledger/Babbage/Rules/Utxo.hs [hunk 3]
```diff
@@ -110,6 +113,9 @@ data BabbageUtxoPredFailure era
     -- together with the minimum value for the given output.
     BabbageOutputTooSmallUTxO
       ![(TxOut era, Coin)]
+  | -- | TxIns that appear in both inputs and reference inputs
+    BabbageNonDisjointRefInputs
+      !(Set (TxIn (EraCrypto era)))
```
Comment (lehins) [inline eras/babbage/impl/src/Cardano/Ledger/Babbage/Rules/Utxo.hs]: Lets add `BabbageNonDisjointRefInputs` to the changelog and switch to `NonEmpty`
```suggestion
    BabbageNonDisjointRefInputs
      !(NonEmpty (TxIn (EraCrypto era)))
```

**Review by lehins — APPROVED** [review]
Looks great.

File: eras/babbage/impl/src/Cardano/Ledger/Babbage/Rules/Utxo.hs (hunk 2 above)
Comment (TimSheard) [reply eras/babbage/impl/src/Cardano/Ledger/Babbage/Rules/Utxo.hs]: done

File: CHANGELOG.md [hunk 4]
```diff
@@ -9,6 +9,7 @@ If you are looking for the Ledger Releasing and Versioning Process then you can
 
 ## 8.8
 
+- Add the constructor BabbageNonDisjointRefInputs to BabbageUtxoPredFailure
```
Comment (lehins) [inline CHANGELOG.md]: Same here wrong changelog.

This should go into `cardano-ledger-babbage` changelog

Comment (TimSheard) [reply CHANGELOG.md]: done

File: eras/babbage/impl/src/Cardano/Ledger/Babbage/Rules/Utxo.hs [hunk 5]
```diff
@@ -194,6 +202,24 @@ feesOK pp tx (UTxO utxo) =
...
+  Set (TxIn (EraCrypto era)) ->
+  Set (TxIn (EraCrypto era)) ->
+  Test (BabbageUtxoPredFailure era)
+disjointRefInputs pp inputs refInputs =
+  when
+    (pvMajor (pp ^. ppProtocolVersionL) > eraProtVerHigh @(BabbageEra (EraCrypto era)))
+    ( case nonEmpty (Set.toList common) of -- Nothing if common is empty
+        Just ne -> failureIf (not (Set.null common)) (BabbageNonDisjointRefInputs ne)
+        Nothing -> pure ()
+    )
```
Comment (lehins) [inline eras/babbage/impl/src/Cardano/Ledger/Babbage/Rules/Utxo.hs]: We now have `failureOnNonEmpty` that will be more efficient and will not force the computation when validation is turned off:


```suggestion
    (failureOnNonEmpty common BabbageNonDisjointRefInputs)
```

See the definition: https://github.com/IntersectMBO/cardano-ledger/blob/b05ba3609910c5e54ea0a27acf93dcd9a5dcd223/libs/small-steps/src/Control/State/Transition/Extended.hs#L413-L414

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Examples/BabbageFeatures.hs [hunk 6]
```diff
@@ -51,12 +49,14 @@ import Cardano.Ledger.Babbage.TxInfo (
  )
 import Cardano.Ledger.BaseTypes (
   Network (..),
+  SlotNo (..),
   StrictMaybe (..),
   mkTxIx,
   mkTxIxPartial,
-  natVersion,
+  -- Inject,
```
Comment (lehins) [inline libs/cardano-ledger-test/src/Test/Cardano/Ledger/Examples/BabbageFeatures.hs]: Dead code
```suggestion
```

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Examples/BabbageFeatures.hs [hunk 7]
```diff
@@ -1103,6 +1139,7 @@ utxoFromTestCaseData pf (TestCaseData txBody' (InitOutputs ofInputs' ofRefInputs
       newTxInOuts = newTxIns `zip` toList (getOutputs pf txBody')
 
       initUtxo = UTxO $ Map.fromList (inputs' ++ refInputs' ++ collateral')
+      -- initUtxo = UTxO $ Map.fromist
```
Comment (lehins) [inline libs/cardano-ledger-test/src/Test/Cardano/Ledger/Examples/BabbageFeatures.hs]: Dead code
```suggestion
```

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Examples/BabbageFeatures.hs [hunk 8]
```diff
@@ -1352,11 +1394,51 @@ genericBabbageFailures pf =
         ]
     ]
 
+-- | Some versions of Data.List.NonEmpty do not export singleton.
+nonEmptySingleton :: a -> NonEmpty a
+nonEmptySingleton x = case nonEmpty [x] of
+  Just y -> y
+  Nothing -> error "Impossible in nonEmptySingleton"
+
```
Comment (lehins) [inline libs/cardano-ledger-test/src/Test/Cardano/Ledger/Examples/BabbageFeatures.hs]: We've discussed this. You can just use `pure`
```suggestion
```

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Examples/BabbageFeatures.hs [hunk 9]
```diff
@@ -1352,11 +1394,51 @@ genericBabbageFailures pf =
...
   testGroup
     "Babbage Features"
     [ genericBabbageFeatures Babbage
     , genericBabbageFailures Babbage
-    -- genericBabbageFeatures Conway -- TODO
+    , genericBabbageFeatures Conway
+    , genericBabbageFailures Conway
+    , testCase "inputs and refinputs overlap in Babbage and don't Fail" $ testExpectSuccessValid Babbage (commonReferenceScript Babbage)
+    , testCase "inputs and refinputs overlap in Conway and Fail" $
+        testExpectUTXOFailure Conway (commonReferenceScript Conway) (BabbageNonDisjointRefInputs (nonEmptySingleton commonTxIn))
```
Comment (lehins) [inline libs/cardano-ledger-test/src/Test/Cardano/Ledger/Examples/BabbageFeatures.hs]:
```suggestion
        testExpectUTXOFailure Conway (commonReferenceScript Conway) (BabbageNonDisjointRefInputs (pure commonTxIn))
```

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Examples/BabbageFeatures.hs [hunk 10]
```diff
@@ -1352,11 +1394,51 @@ genericBabbageFailures pf =
...
+        (UTXO pf)
+        env
+        state
+        tx'
+        ( \x -> case x of
+            Left [predfail] -> assertEqual "unexpected failure" predfail failure
+            Left _ -> assertFailure "not exactly one failure"
+            Right _ -> assertFailure "testExpectUTXOFailure succeeds"
+        )
+testExpectUTXOFailure _ _ _ = error ("testExpectUTXOFailure is only good in Conway Era")
```
Comment (lehins) [inline libs/cardano-ledger-test/src/Test/Cardano/Ledger/Examples/BabbageFeatures.hs]: Redundant parenthesis
```suggestion
testExpectUTXOFailure _ _ _ = error "testExpectUTXOFailure is only good in Conway Era"
```

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Examples/BabbageFeatures.hs [hunk 11]
```diff
@@ -1352,11 +1394,51 @@ genericBabbageFailures pf =
...
+  , Reflect era
+  , BabbageEraTxBody era
+  ) =>
+  Proof era ->
+  TestCaseData era ->
+  PredicateFailure (EraRule "UTXO" era) ->
+  Assertion
+testExpectUTXOFailure pf@Conway tc failure =
+  let tx' = txFromTestCaseData pf tc
+      (InitUtxo inputs' refInputs' collateral') = initUtxoFromTestCaseData pf tc
```
Comment (lehins) [inline libs/cardano-ledger-test/src/Test/Cardano/Ledger/Examples/BabbageFeatures.hs]: Redundant parenthesis
```suggestion
      InitUtxo inputs' refInputs' collateral' = initUtxoFromTestCaseData pf tc
```

**Pull request merged**